image/preseed/reset.go

```diff
@@ -31,6 +31,50 @@ import (
 	apparmor_sandbox "github.com/snapcore/snapd/sandbox/apparmor"
 )
 
+// bash-completion symlinks; note there are symlinks that point at
+// completer, and symlinks that point at the completer symlinks.
+// e.g.
+// lxd.lxc -> /snap/core/current/usr/lib/snapd/complete.sh
+// lxc -> lxd.lxc
+func resetCompletionSymlinks(completersPath string, preseedChroot string) error {
+	files, err := ioutil.ReadDir(filepath.Join(preseedChroot, completersPath))
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("error reading %s: %v", completersPath, err)
+	}
+	completeShSymlinks := make(map[string]string)
+	var otherSymlinks []string
+
+	// pass 1: find all symlinks pointing at complete.sh
+	for _, fileInfo := range files {
+		if fileInfo.Mode()&os.ModeSymlink == 0 {
+			continue
+		}
+		fullPath := filepath.Join(preseedChroot, completersPath, fileInfo.Name())
+		if dirs.IsCompleteShSymlink(fullPath) {
+			if err := os.Remove(fullPath); err != nil {
+				return fmt.Errorf("error removing symlink %s: %v", fullPath, err)
+			}
+			completeShSymlinks[fileInfo.Name()] = fullPath
+		} else {
+			otherSymlinks = append(otherSymlinks, fullPath)
+		}
+	}
+	// pass 2: find all symlinks that point at the symlinks found in pass 1.
+	for _, other := range otherSymlinks {
+		target, err := os.Readlink(other)
+		if err != nil {
+			return fmt.Errorf("error reading symlink target of %s: %v", other, err)
+		}
+		if _, ok := completeShSymlinks[target]; ok {
+			if err := os.Remove(other); err != nil {
+				return fmt.Errorf("error removing symlink %s: %v", other, err)
+			}
+		}
```

Comment:
Just wondering: would it make sense to just check if the target exists, and remove the current symlink if the target does not exist?
It would be slightly less efficient, but then we wouldn't need to worry about symlinks like `"./file-in-same-dir"`. Though I'm probably inventing problems here. :-)

Reply:
This is not my code. This was moved for this PR, but this part was not modified. I would prefer not to change it.

The way symlinks are created is:
 - one layer of symlinks point to some `complete.sh` from somewhere else (within snapd snap most of the time).
 - for every alias, there is a symlink pointing to the main symlink.

So to me this code looks like a correct way to remove symlinks created by snapd.

I am not totally sure if that applies to preseeding, but I do not think we should remove invalid symlinks not created by snapd.